plugin/src/main/java/org/opensearch/ml/cluster/MLSyncUpCron.java

```diff
@@ -91,6 +129,171 @@ public void run() {
+                        loadingModels,
+                        modelWorkerNodes,
+                        modelId,
+                        state,
+                        lastUpdateTime,
+                        planningWorkerNodeCount,
+                        currentWorkerNodeCountInIndex
+                    );
+                    if (mlModelState != null) {
+                        newModelStates.put(modelId, mlModelState);
+                    }
+                }
+                bulkUpdateModelState(modelWorkerNodes, newModelStates);
+            }, e -> {
+                updateModelStateSemaphore.release();
+                log.error("Failed to search models", e);
+            }));
+        } catch (Exception e) {
+            updateModelStateSemaphore.release();
+            log.error("Failed to refresh model state", e);
+        }
+    }
+
+    private MLModelState getNewModelState(
+        Map<String, Set<String>> loadingModels,
+        Map<String, Set<String>> modelWorkerNodes,
+
+        String modelId,
+        MLModelState state,
+        Long lastUpdateTime,
+        int planningWorkerNodeCount,
+        int currentWorkerNodeCountInIndex
+    ) {
+        Set<String> loadTaskNodes = loadingModels.get(modelId);
+        if (loadTaskNodes != null && loadTaskNodes.size() > 0 && state != MLModelState.LOADING) {
+            // no
+            return MLModelState.LOADING;
+        }
+        int currentWorkerNodeCount = modelWorkerNodes.containsKey(modelId) ? modelWorkerNodes.get(modelId).size() : 0;
+        if (currentWorkerNodeCount == 0
+            && state != MLModelState.LOAD_FAILED
+            && !(state == MLModelState.LOADING
+                && lastUpdateTime != null
+                && lastUpdateTime + LOAD_MODEL_TASK_GRACE_TIME_IN_MS > Instant.now().toEpochMilli())) {
+            // If model not deployed to any node and no node is loading the model, then set model state as LOAD_FAILED
+            return MLModelState.LOAD_FAILED;
+        }
+        if (currentWorkerNodeCount > 0) {
+            if (currentWorkerNodeCount < planningWorkerNodeCount
+                && (state != MLModelState.PARTIALLY_LOADED || currentWorkerNodeCountInIndex != currentWorkerNodeCount)) {
+                // If model deployed to some node/nodes, but not deployed to all nodes planned by user,
+                // then set model state as PARTIALLY_LOADED.
+                return MLModelState.PARTIALLY_LOADED;
+            } else if (planningWorkerNodeCount > 0 && currentWorkerNodeCount >= planningWorkerNodeCount && state != MLModelState.LOADED) {
+                if (currentWorkerNodeCount > planningWorkerNodeCount) {
+                    // This case should not happen that model loaded to more nodes than planned. So log this as warning if
+                    // it happens.
+                    log
+                        .warn(
+                            "Model {} loaded on more nodes [{}] than planing worker node[{}]",
```

Comment:
[nit] than planing worker -> than planning worker

Reply:
Just curious to know how this warning is useful to customers. What do we expect user to do if they see this warning?

Reply:
That case is not expected to happen. But if user found this edge case happens, this error can give some clue like when this error happened. Then we can analyze other log like when user started loading model, if user re-loaded model etc to find out the root cause.